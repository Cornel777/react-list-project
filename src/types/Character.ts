export type Gender = 'male' | 'female' | 'n/a' | 'none';
export type Color = string;
export type URL = string;

export interface Character {
  name: string;
  height: string;
  mass: string;
  hair_color: Color;
  skin_color: Color;
  eye_color: Color;
  birth_year: string;
  gender: Gender;
  homeworld: URL;
  films: URL[];
  species: URL[];
  vehicles: URL[];
  starships: URL[];
  created: string;
  edited: string;
  url: URL;
}
