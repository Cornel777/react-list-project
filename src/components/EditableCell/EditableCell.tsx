import { useState, useRef } from 'react';
import { TextField, IconButton } from '@mui/material';
import ModeEditIcon from '@mui/icons-material/ModeEdit';
import { StyledTableCell } from "../CharacterList/styled/styledListTable";
import { Character } from "../../types/Character";

interface EditableCellProps {
  value: string;
  field: 'height' | 'gender';
  favorite: Character;
  isFocused: boolean;
  onSave: (favorite: Character, value: string, field: string) => void;
}

const EditableCell: React.FC<EditableCellProps> = ({ 
  value, 
  field, 
  favorite, 
  isFocused,
  onSave 
}) => {
  const [isEditing, setIsEditing] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);

  const handleEdit = () => setIsEditing(true);
  
  const handleSave = (newValue: string) => {
    onSave(favorite, newValue, field);
    setIsEditing(false);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && inputRef.current) {
      handleSave(inputRef.current.value);
    }
    if (e.key === 'Escape') {
      setIsEditing(false);
    }
  };

  return (
    <StyledTableCell
    >
      {isEditing ? (
        <TextField 
          inputRef={inputRef}
          defaultValue={value}
          onBlur={(e) => handleSave(e.target.value)}
          onKeyDown={handleKeyDown}
          autoFocus
          size="small"
        />
      ) : (
        <>
          {value}
          {isFocused && (
            <IconButton
              aria-label={`edit ${field}`}
              sx={{ marginLeft: '0.5rem' }}
              onClick={handleEdit}
            >
              <ModeEditIcon />
            </IconButton>
          )}
        </>
      )}
    </StyledTableCell>
  );
};

export default EditableCell;