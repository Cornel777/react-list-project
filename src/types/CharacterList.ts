import { Character } from "./Character";
import { URL } from "./Character";

export interface CharacterList {
  count: number;
  next: URL | null;
  previous: URL | null;
  results: Character[];
}