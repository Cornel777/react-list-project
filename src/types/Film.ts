import { URL } from './Character';

export interface Film {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  characters: URL[];
  planets: URL[];
  starships: URL[];
  vehicles: URL[];
  species: URL[];
  created: Date;
  edited: Date;
  url: URL;
}