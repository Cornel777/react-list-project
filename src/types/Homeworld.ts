import { URL } from "./Character";

export interface Homeworld {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: URL[];
  created: string;
  edited: string;
  url: URL;
}

