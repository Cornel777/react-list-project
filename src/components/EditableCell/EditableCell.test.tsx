import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import EditableCell from './EditableCell';
import { Character } from '../../types/Character';


describe('EditableCell', () => {
  const defaultProps = {
    value: '172',
    field: 'height' as const,
    favorite: {
      name: 'Luke',
      height: '172',
      gender: 'male',
      url: 'https://swapi.dev/api/people/1/',
      homeworld: '',
      films: [],
      species: [],
      vehicles: [],
      starships: [],
      created: '',
      edited: '',
      birth_year: '',
      eye_color: '',
      hair_color: '',
      skin_color: '',
      mass: ''
    } as Character,
    isFocused: false,
    onSave: vi.fn()
  };

  it('matches the snapshot', () => {
    const { container } = render(<EditableCell {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the initial value', () => {
    render(<EditableCell {...defaultProps} />);
    expect(screen.getByText('172')).toBeInTheDocument();
  });

  it('renders the edit button', () => {
    render(<EditableCell {...defaultProps} isFocused/>);
    expect(screen.getByLabelText('edit height')).toBeInTheDocument();
  });

  it('enters edit mode when the edit button is clicked', () => {
    render(<EditableCell {...defaultProps} isFocused/>);
    fireEvent.click(screen.getByLabelText('edit height'));
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('saves the edits when the enter key is pressed', () => {
    render(<EditableCell {...defaultProps} isFocused/>);
    fireEvent.click(screen.getByLabelText('edit height'));
    fireEvent.change(screen.getByRole('textbox'), { target: { value: '180' } });
    fireEvent.keyDown(screen.getByRole('textbox'), { key: 'Enter' });

    expect(defaultProps.onSave).toHaveBeenCalledWith(defaultProps.favorite, '180', 'height');
  });

  it('cancels the edits when the escape key is pressed', () => {
    render(<EditableCell {...defaultProps} isFocused/>);
    fireEvent.click(screen.getByLabelText('edit height'));
    fireEvent.change(screen.getByRole('textbox'), { target: { value: '180' } });
    fireEvent.keyDown(screen.getByRole('textbox'), { key: 'Escape' });

    expect(screen.getByText('172')).toBeInTheDocument();
  });

});