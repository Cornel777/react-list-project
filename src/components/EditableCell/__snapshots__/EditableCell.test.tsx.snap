// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableCell > matches the snapshot 1`] = `
<div>
  <td
    class="MuiTableCell-root MuiTableCell-sizeMedium css-1cutd21-MuiTableCell-root"
  >
    172
  </td>
</div>
`;
